### DISCUSSION TITLE
About who should take care of `HostPhysAddr` and `GuestPhysAddr`

### DISCUSSION THREAD

**Discussion by hky1999** [arceos-hypervisor/axvisor#8] [General]
From the perspective of an ordinary OS address space, we only need to consider the concepts of physical address and virtual address.
That is, `VirtAddr` and `PhysAddr` from [memory_addr](https://crates.io/crates/memory_addr) crate.

However, as a hypervisor, we need to take care of two-stage address translation, which brings the concept of `GuestVirtAddr` and `GuestPhysAddr`, along with `HostVirtAddr` and `HostPhysAddr`.

I am  currently working on refactor the address space management for guest VM in `axvm` crate.

I find that [axvcpu](https://github.com/arceos-hypervisor/axvcpu/blob/main/src/lib.rs) has definition of `GuestVirtAddr` and `GuestPhysAddr`, like this:

```Rust
/// Guest virtual address.
pub type GuestVirtAddr = usize;
/// Guest physical address.
pub type GuestPhysAddr = usize;
```

Also, [axvm](https://github.com/arceos-hypervisor/arceos-umhv/blob/master/crates/axvm/src/mm/mod.rs) also has its own definition of `GuestVirtAddr` and `GuestPhysAddr` (althrough it simply regards them as `usize`).

```Rust
/// Guest virtual address.
pub type GuestVirtAddr = usize;
/// Guest physical address.
pub type GuestPhysAddr = usize;
/// Host virtual address.
pub type HostVirtAddr = VirtAddr;
/// Host physical address.
pub type HostPhysAddr = PhysAddr;
```

During my refactor process, I hope to integrate them into the `axaddrspace` crate, like this:

```Rust
use memory_addr::{PhysAddr, VirtAddr};

/// Guest virtual address.
pub type GuestVirtAddr = VirtAddr;
/// Guest physical address.
pub type GuestPhysAddr = VirtAddr;
/// Host virtual address.
pub type HostVirtAddr = VirtAddr;
/// Host physical address.
pub type HostPhysAddr = PhysAddr;
```
Because `axaddrspace` essentially needs to manage the translation between `GuestPhysAddr` to `HostPhysAddr`, which is `VirtAddr` to `PhysAddr` from the aspect of `memory_addr` crate.

My question is, from the perspective of `arceos-vmm` app, does it need to take care of `HostPhysAddr` and `GuestPhysAddr`, or does it only need to take care of the `VirtAddr` and `PhysAddr` perspective?

Note: we need to implement `AxVMHalImpl` for `axvm`'s `AxVMHal Trait` in `arceos-vmm` app.

I just use `memory_addr::{PhysAddr, VirtAddr}` in arceos-vmm because I think only `axvm` crate should care for  `HostPhysAddr` and `GuestPhysAddr`.

**Comment by hky1999** [comment]
@aarkegz

**Reply by aarkegz** [reply]
reading……

**Comment by aarkegz** [comment]
An `axaddrspace` is undoubtedly necessary.

> During my refactor process, I hope to integrate them into the `axaddrspace` crate, like this:
> 
> ```rust
> use memory_addr::{PhysAddr, VirtAddr};
> 
> /// Guest virtual address.
> pub type GuestVirtAddr = VirtAddr;
> /// Guest physical address.
> pub type GuestPhysAddr = VirtAddr;
> /// Host virtual address.
> pub type HostVirtAddr = VirtAddr;
> /// Host physical address.
> pub type HostPhysAddr = PhysAddr;
> ```

⬆️Why is `GuestPhysAddr` `VirtAddr`?

There's one more thing: in rust, type alias does not create new types. Aliasing both `GuestVirtAddr` and `HostVirtAddr` to `VirtAddr` means they can be used interchangeably, which sounds a little bit weird.

**Reply by hky1999** [reply]
Yep, even `GuestPhysAddr` and `HostVirtAddr` can be used interchangeably.

`GuestPhysAddr` is regared as `VirtAddr` because we want to reuse `page_table_multiarch` , which  only accepts parameters of type `VirtAddr` as the start of virtual page to be mapped.

In fact, it's the translation between `GuestPhysAddr` (as `VirtAddr`) and `HostPhysAddr`  (as `PhysAddr`) that matters.

Maybe eventually, we still need to add two new wrappers for guest memory address?

**Reply by hky1999** [reply]
ref of page_table_multiarch: https://github.com/arceos-org/page_table_multiarch/blob/main/page_table_multiarch/src/bits64.rs#L61

**Reply by aarkegz** [reply]
I agree that GPA is more like HVA instead of HPA. But GPA is not HVA, and some methods of `VirtAddr` like `as_ptr` are very dangerous for GPA.

> Maybe eventually, we still need to add two new wrappers for guest memory address?

Yes, very likely.

> `GuestPhysAddr` is regared as `VirtAddr` because we want to reuse `page_table_multiarch` , which only accepts parameters of type `VirtAddr` as the start of virtual page to be mapped.

Maybe `page_table_multiarch` should make parameter types more generic.

**Reply by hky1999** [reply]
I agree.  @equation314

**Reply by equation314** [reply]
Some things like this? 
https://github.com/rcore-os/RVM1.5/blob/main/src/memory/paging.rs

**Reply by hky1999** [reply]
Exactly, I think it is missed in current [page_table_multiarch/bits64](https://github.com/arceos-org/page_table_multiarch/blob/main/page_table_multiarch/src/bits64.rs#L61) ?

**Reply by aarkegz** [reply]
> Some things like this? 
> 
> https://github.com/rcore-os/RVM1.5/blob/main/src/memory/paging.rs

Yes, `Into<usize> + Sized` should be enough.

**Reply by equation314** [reply]
Writing too many generics is disgusting...

**Reply by aarkegz** [reply]
> Writing too many generics is disgusting...

That's an inevitable cost of abstraction🙃

**Reply by aarkegz** [reply]
Maybe it can be made an associated type of `PTE`?

**Reply by equation314** [reply]
Ok, it may appear in the next release. Anyone wanna do this?

**Reply by aarkegz** [reply]
> Maybe it can be made an associated type of `PTE`?

At least no more generic parameter s for the page table.

**Reply by aarkegz** [reply]
> Ok, it may appear in the next release. Anyone wanna do this?

I can implement this and axaddrspace

**Reply by hky1999** [reply]
> > Ok, it may appear in the next release. Anyone wanna do this?
> 
> I can implement this and axaddrspace

`axaddrspace` has been implemented and will be published soon.

**Reply by aarkegz** [reply]
> > > Ok, it may appear in the next release. Anyone wanna do this?
> 
> > 
> 
> > I can implement this and axaddrspace
> 
> 
> 
> `axaddrspace` has been implemented and will be published soon.
> 
> 

Oh, that's nice

**Reply by equation314** [reply]
We just need to add an association `VA` type in `PagingMetaData`.

**Reply by hky1999** [reply]
> Maybe it can be made an associated type of `PTE`?

@aarkegz  You can improve this,

**Reply by aarkegz** [reply]
> We just need to add an association `VA` type in `PagingMetaData`.

🫡👌

**Reply by hky1999** [reply]
> > > Ok, it may appear in the next release. Anyone wanna do this?
> > 
> > 
> > I can implement this and axaddrspace
> 
> `axaddrspace` has been implemented and will be published soon.

[axaddressspace](https://github.com/arceos-hypervisor/axaddrspace) has been published, currently only works in x86_64 is tested.

I will open a new discussion soon about questions related to memory management.

**Comment by aarkegz** [comment]
Updates allowing generic virtual address have been merged into `page_table_multiarch`. I think it's time to update `axaddrspace` to introduce new GPA and GVA types. @hky1999

**Reply by hky1999** [reply]
Yes, I am working on it, sorry for delay

**Reply by hky1999** [reply]
> Updates allowing generic virtual address have been merged into `page_table_multiarch`. I think it's time to update `axaddrspace` to introduce new GPA and GVA types. @hky1999

This is more complicated than we throught.

The same modification we did to `page_table_multiarch`  (adding an association VA type) should be deployed in both `memory_set`, for example:

```Rust
// memory_set/src/area.rs
pub trait MappingBackend<F: Copy, P>: Clone {
    /// What to do when mapping a region within the area with the given flags.
    fn map(&self, start: VirtAddr, size: usize, flags: F, page_table: &mut P) -> bool;
    /// What to do when unmaping a memory region within the area.
    fn unmap(&self, start: VirtAddr, size: usize, page_table: &mut P) -> bool;
}
```

**Reply by hky1999** [reply]
The complexity:

in `page_table_multiarch`, `VA` type in `PagingMetaData` can be easily covered by

`type VirtAddr: Into<usize> + From<usize> + Copy;`

however, in `memory_set`, `VirtAddr`'s `as_usize()` is used almost everywhere, it's a const function which makes it impossible to use a trait to describe the behavior of `VirtAddr`.

The problem is, how can we make parameter types like `memory_addr::VirtAddr` in `memory_set` more generic, so `memort_set` can be used in hypervisor's guest address space management.

**Reply by hky1999** [reply]
related PR:
arceos-umhv: https://github.com/arceos-hypervisor/arceos-umhv/pull/15
axaddrspace: https://github.com/arceos-hypervisor/axaddrspace/pull/3

@aarkegz @equation314
